File: 733-flood-fill/733-flood-fill.rb
# @param {Integer[][]} image
# @param {Integer} sr
# @param {Integer} sc
# @param {Integer} color
# @return {Integer[][]}
def out_of_bounds(image, r, c)
    row_bounds = [-1, image.length]
    col_bounds = [-1, image[0].length]
    return true if row_bounds.include?(r) || col_bounds.include?(c)
    false
end

def dfs(image, r, c, old_color, new_color, visit)
    return if out_of_bounds(image, r, c) || image[r][c] != old_color || visit.include?([r, c])
    visit << [r, c]
    image[r][c] = new_color
    
    [[1, 0], [-1, 0], [0, 1], [0, -1]].each do |dr, dc|
         dfs(image, r + dr, c + dc, old_color, new_color, visit)
    end
end

def flood_fill(image, sr, sc, color)
    visit = Set.new
    dfs(image, sr, sc, image[sr][sc], color, visit)
    image
end

File: 733-flood-fill/733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        ROWS, COLS = len(image), len(image[0])
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        visit = set()
        original_color = image[sr][sc]
        
        def out_of_bounds(r, c):
            return r < 0 or c < 0 or r == ROWS or c == COLS
        
        def dfs(r, c):
            if out_of_bounds(r, c) or (r, c) in visit or image[r][c] != original_color:
                return
            
            visit.add((r, c))
            image[r][c] = color
            for dr, dc in directions:
                dfs(r + dr, c + dc)
                
        dfs(sr, sc)
        return image